working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2159,17 +2209,233 @@ To match a pattern `p` against a value `v`:
+`args`.
+
+Given a set of patterns `s` matching a value expression `v`, we bind an
+invocation key to each member invocation and record field access in `s` like so:
+
+1.  Let `i` be an invocation key with no parent, no extension type, named
+    `this`, with an empty argument list. *This is the root node of the
+    invocation key tree and represents the matched value itself.*
+
+2.  For each pattern `p` in `s` with parent invocation `i`, bind invocation keys
+    to it and its subpatterns using the following procedure:
+
+To bind invocation keys in a pattern `p` using parent invocation `i`:
+
+*   **Logical-or** or **logical-and**:
+
+    1.  Bind invocations in the left and right subpatterns using target `i`.
+
+*   **Relational**:
+
+    1.  Resolve the relational operator to a member.
+
+    2.  If it resolves to an instance method, then bind `i : (op, [arg])` to
+        the operator method invocation where `op` is the name of the operator
+        and `arg` is the right operand value.
+
+    3.  Else it resolves to an extension method. Bind `i : type(op, [arg])` to
+        the operator method invocation where `type` is the type the extension
+        method resolved to, `op` is the name of the operator and `arg` is the
+        right operand value.
+
+*   **Cast**, **null-check**, **null-assert**, or **parenthesized**:
+
+    1.  Bind invocations in the subpattern using target `i`.
+
+*   **Constant**:
+
+    1.  Bind `i : ("==", [arg])` to the `==` method invocation where`arg` is the
+        right operand value.
+
+*   **Variable**:
+
+    1.  Nothing to do.
+
+*   **List**:
+
+    1.  Bind `i : ("length", [])` to the `length` getter invocation.
+
+    2.  For each element subpattern:
+
+        1.  Let `e` be `i : ("[]", [index])` where `index` is the zero-based
+            index of this element subpattern.
+
+        2.  Bind `e` to the `[]` invocation for this element.
+
+        3.  Bind invocations in the element subpattern using target `e`.
+
+*   **Map**:
+
+    1.  Bind `i : ("length", [])` to the `length` getter invocation.
+
+    2.  For each entry in `p`:
+
+        1.  Bind `i : ("containsKey()", [key])` to the `containsKey()`
+            invocation where `key` is entry's key constant value.
+
+        2.  Let `e` be `i : ("[]", [key])` where `key` is entry's key constant
+            value.
+
+        3.  Bind `e` to the `[]` invocation for this entry.
+
+        4.  Bind invocations in the entry value subpattern using target `e`.
+
+*   **Record**:
+
+    1.  For each field `f` in `p`:
+
+        1.  Let `f` be `i : (field, [])` where `field` is the corresponding
+            getter name for the field.
+
+        2.  Bind `e` to the field accessor for this field.
+
+        3.  Bind invocations in the field subpattern using target `e`.
+
+*   **Extractor**:
+
+    1.  For each field `f` in `p`:
+
+        1.  Resolve the getter name to a member.
```

Comment:
Same comment as above: At least need to say that it resolves against static type, and what it means to resolve against `dynamic`/`Never`

Reply:
Done.